File: 2500/2589.py
from collections import deque
n,m=map(int,input().split())
D=[input() for _ in range(n)]
ans=0
dx,dy=[0,0,1,-1],[1,-1,0,0]
def BFS(i,j):
    global ans
    S=[[-1]*m for _ in range(n)]
    q=deque([(i,j)]);S[i][j]=0
    while q:
        x,y=q.popleft()
        for i in range(4):
            nx,ny=x+dx[i],y+dy[i]
            if 0<=nx<n and 0<=ny<m and D[nx][ny]=='L' and S[nx][ny]==-1:
                q.append((nx,ny));S[nx][ny]=S[x][y]+1;ans=max(ans,S[nx][ny])
for i in range(n):
    for j in range(m):
        if D[i][j]=='L':BFS(i,j)
print(ans)

File: 2500/2589.cpp
#include <bits/stdc++.h>
using namespace std;

int n,m;
char D[51][51];
queue<pair<int,int>>q;
pair<int,int>p;
int ans,nx,ny;
int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};

int BFS(int x, int y){
    int z = 0;
    int check[101][101];
    memset(check,-1,sizeof(check));
    q.push(make_pair(x,y));
    check[x][y]=0;
    while (!q.empty()){
        p = q.front();q.pop();
        x = p.first;
        y = p.second;
        for(int i = 0; i<4; i++){
            nx = x+dx[i];
            ny = y+dy[i];
            if (0<=nx && nx<n && 0<=ny && ny<m && D[nx][ny]=='L' && check[nx][ny]==-1){
                z=max(z,check[x][y]+1); check[nx][ny]=check[x][y]+1;q.push(make_pair(nx,ny));
            }
        }
    }
    return z;
}

int main(){
    cin >> n >> m;
    ans=0;
    for(int i = 0; i < n; i++){
        scanf("%s",&D[i]);
    }
    for(int i = 0; i < n; i++){
        for(int j = 0; j< m; j++){
            if(D[i][j]=='L') ans=max(ans,BFS(i,j));
        }
    }
    cout << ans << endl;
}